src/data/sym/card.lean

```diff
@@ -5,41 +5,212 @@ Authors: Yaël Dillies, Bhavik Mehta
+def encode [decidable_eq α] (s : sym (option α) n.succ) : sym α n.succ ⊕ sym (option α) n :=
+if h : none ∈ s
+then inr (s.erase none h)
+else inl (s.attach.map $ λ o,
+  option.get $ option.ne_none_iff_is_some.1 $ ne_of_mem_of_not_mem o.2 h)
+
+/-- From the output of `encode`, the `decode` function reconstructs the original input. If the
+output contains `n + 1` elements, the original input can be reconstructed by casting `α` back
+to `option α`. Otherwise, an instance of `none` has been removed and the input can be
+reconstructed by adding it back. -/
+def decode : sym α n.succ ⊕ sym (option α) n → sym (option α) n.succ
+| (inl s) := s.map embedding.coe_option
+| (inr s) := none :: s
+
+variables {α n}
+
+@[simp] lemma decode_inl (s : sym α n.succ) : decode α n (inl s) = s.map embedding.coe_option := rfl
+@[simp] lemma decode_inr (s : sym (option α) n) : decode α n (inr s) = none :: s := rfl
+
+variables (α n)
+
+/-- As `encode` and `decode` are inverses of each other, `sym (option α) n.succ` is equivalent
+to `sym α n.succ ⊕ sym (option α) n`. -/
+def option_succ_equiv [decidable_eq α] : sym (option α) n.succ ≃ sym α n.succ ⊕ sym (option α) n :=
+{ to_fun := encode α n,
+  inv_fun := decode α n,
+  left_inv := λ s, begin
+    unfold encode,
+    split_ifs,
+    { exact cons_erase _ },
+    { simp only [decode, map_map, subtype.mk.inj_eq, comp, option.coe_get,
+        embedding.coe_option_apply, subtype.val_eq_coe, map_map],
+      convert s.attach_map_coe, },
+  end,
+  right_inv := begin
+    rintro (s | s),
+    { unfold encode,
+      split_ifs,
+      { obtain ⟨a, _, ha⟩ := multiset.mem_map.mp h,
+        exact option.some_ne_none _ ha },
+      { refine map_injective (option.some_injective _) _ _,
+        convert eq.trans _ (decode α n (inl s)).attach_map_coe,
+        simp } },
+    { exact (dif_pos $ mem_cons_self _ _).trans (congr_arg _ $ erase_cons_head s _ _) }
+  end }
+
+end
+
+/-- `multichoose n k` is the number of multisets of cardinality `k` from a type of cardinality `n`.
+That is, it's the number of ways to select `k` items (up to permutation) from `n` items
+with replacement.
+
+This is defined as `nat.choose (n + k - 1) k`. It is related to the cardinality of `sym` in
+`sym.multichoose_eq`. -/
+def multichoose (n k : ℕ) := (n + k - 1).choose k
+
+lemma card_sym_rec (n : ℕ)
+  [fintype (sym (option α) n.succ)] [fintype (sym α n.succ)] [fintype (sym (option α) n)] :
+  card (sym (option α) n.succ) = card (sym α n.succ) + card (sym (option α) n) :=
+by { classical, simpa only [card_sum.symm] using fintype.card_congr (option_succ_equiv α n) }
+
+lemma multichoose_rec (n k : ℕ) :
+  multichoose n.succ k.succ = multichoose n k.succ + multichoose n.succ k :=
+by simp [multichoose, nat.choose_succ_succ, nat.add_comm, nat.add_succ]
+
+/--
+Inhabited types are equivalent to `option β` for some `β` by identifying `default α` with `none`.
+-/
+def {u} sigma_equiv_option_of_nonempty (α : Type u) [inhabited α] [decidable_eq α] :
```

Comment:
This isn't `symm`-specific and could move elsewhere.

Reply:
I've moved this to `data/finset/basic`